pytorch_lightning/trainer/trainer.py

```diff
@@ -2026,6 +2026,23 @@ def should_rank_save_checkpoint(self) -> bool:
     def num_nodes(self) -> int:
         return getattr(self.strategy, "num_nodes", 1)
 
+    @property
+    def device_ids(self) -> List[int]:
+        """List of device indexes per node."""
+        devices = getattr(self.strategy, "parallel_devices", [self.strategy.root_device])
+        device_ids = []
+        for idx, device in enumerate(devices):
+            if isinstance(device, torch.device):
+                device_ids.append(device.index or idx)
+            elif isinstance(device, int):
+                device_ids.append(device)
+        return device_ids
+
+    @property
+    def num_devices(self) -> int:
+        """Number of devices the trainer uses per node."""
+        return len(self.strategy.parallel_devices) if isinstance(self.strategy, ParallelStrategy) else 1
```

Comment:
my assumption is that `num_devices` will be infrequently accessed, so I would prefer we rely on `device_ids` for the length to avoid these falling out of sync. If we need to optimize the performance, we can always do so without it being a breaking change.